src/middlewared/middlewared/plugins/smart.py

```diff
@@ -296,6 +297,101 @@ class Config:
+                    Str('type', enum=['LONG', 'SHORT', 'CONVEYANCE', 'OFFLINE'], required=True),
+                )
+            ]
+        )
+    )
+    async def manual_test(self, disks):
+        """
+        Run manual SMART tests for `disks`.
+
+        `type` indicates what type of SMART test will be ran and must be specified.
+        """
+        verrors = ValidationErrors()
+        if not disks:
+            verrors.add(
+                'manual_test.disks',
+                'Please specify at least one disk.'
+            )
+        else:
+            test_disks_list = []
+            disks_data = await self.middleware.call('disk.query')
+            devices = await camcontrol_list()
+
+            for disk in disks:
+                for d in disks_data:
+                    if disk['identifier'] == d['identifier']:
+                        current_disk = d
+                        test_disks_list.append({
+                            'disk': current_disk['name'],
+                            **disk
+                        })
+                        break
+                else:
+                    verrors.add(
+                        'manual_test.disks.identifier',
+                        f'{disk["identifier"]} is not valid. Please provide a valid disk identifier.'
+                    )
+                    continue
+
+                if current_disk['name'] is None or current_disk['name'].startswith('nvd'):
+                    verrors.add(
+                        'manual_test.disks.identifier',
+                        f'Test cannot be performed for {disk["identifier"]} disk. Failed to retrieve name.'
+                    )
+
+                device = devices.get(current_disk['name'])
+                if not device:
+                    verrors.add(
+                        'manual_test.disks.identifier',
+                        f'Test cannot be performed for {disk["identifier"]}. Unable to retrieve disk details.'
+                    )
+
+        verrors.check()
+
+        return list(
+            await asyncio_map(functools.partial(self.__manual_test, devices), test_disks_list, 16)
+        )
+
+    async def __manual_test(self, devices, disk):
+        device = devices.get(disk['disk'])
+        args = await get_smartctl_args(disk['disk'], device)
+
+        proc = await run(
+            list(
+                filter(bool, ['smartctl', '-t', disk['type'].lower(), '-C' if disk['mode'] == 'FOREGROUND' else None])
+            ) + args,
+            check=False, encoding='utf8'
+        )
+
+        output = {}
+        if proc.returncode:
+            output['error'] = proc.stderr
+            self.middleware.logger.debug(
+                f'Self test for {disk["disk"]} failed with {proc.returncode} return code.'
+            )
+        else:
+            time_details = re.findall('test will complete after(.*)', proc.stdout, re.IGNORECASE)
```

Comment:
Lets compile the PR at module level please, something like RE_TIME_DETAILS = re.compile(r'test will complete after(.*)', re.IGNORECASE)

So its not compiled on every disk/usage.